---
title: "Normalised Counts"
author: "Prasanth Sivakumar"
date: "`r Sys.Date()`"
output: html_document
params:
  counts.matrix: 'counts_matrix',
  sample.table: 'sample_table'
---

Auto-generated document to provide basic quality control steps on the featureCounts counts matrix output. \


Converts the output into a matrix format, normalises by size factors, and displays basic comparisons pre- and post-normalisation. \


Generates a raw counts matrix data file (**raw_counts_matrix.RData**), and a normalised counts matrix data file (**norm_counts_matrix.RData**).

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo=FALSE, warning=FALSE)
```

```{r libs, include=FALSE}
library(tidyverse)
library(DESeq2)
```

Input data is the gene expression tab file output from featureCounts, and the sample table containing sample names from the pipeline input.

```{r data, include=TRUE}
# load counts matrix file and sample table
df <- read.table('C:/Users/Prasanth/Documents/barcella_sepsis/gene_expression/barcella_sepsis_gene_counts_old.tab', header = T, row.names = 1, skip = 1)
st <- read.table('C:/Users/Prasanth/Documents/barcella_sepsis/barcella_sepsis_sample_table.tab')
```

```{r cm}
# convert counts matrix file to counts matrix
cm <- df %>%
  dplyr::select(-c(1:5)) %>%
  apply(c(1, 2), as.integer) %>%
  as.matrix()
```

Dataframe of name conversion. \
'**counts_matrix_name**' refers to the sample names in the tab file output from featureCounts - refers to the full path of the input BAM files. \
'**new_name**' refers to the new sample names assigned to the counts matrix, extracted from the paths. \
Check that the order of the old and new names match equivalent samples.

```{r ids}
# set names for sample table columns
names(st) <- c('name', 'fq1', 'fq2', 'cond')

# convert counts matrix path headers to IDs only
ids <- str_remove(colnames(cm), '^.*bams.') %>% str_remove('_Aligned.*$')

# create comparison dataframe of old and new matrix IDS
names.df <- data.frame(
  counts_matrix_name = colnames(cm), 
  new_names = ids
)
print(names.df)

# set new matrix IDs
colnames(cm) <- ids
```

Dimension of the counts matrix before filtering out genes with 0 counts in any one sample: rows (genes) by columns (samples).

```{r dims}
# remove all genes wiht 0 counts in any sample from matrix
filtZeroCounts <- function(mat){
  mat[apply(mat, 1, function(x) all(x !=0 )), ]
}

print(dim(cm))
```

Dimensions of the counts matrix after filtering.

```{r zero}
cm <- filtZeroCounts(cm)
print(dim(cm))
```

Bar chart of total counts assigned to genes in each sample, before normalisation. 

```{r sum}
sf <- estimateSizeFactorsForMatrix(cm)

plotColsums <- function(mat){ 
  colSums(mat) %>%
    as.data.frame() %>% 
    `colnames<-`('sf') %>% 
    rownames_to_column('sample') %>% 
    ggplot(aes(sample, sf)) + 
    geom_col() +
    theme_classic() +
    theme(axis.text.x = element_text(angle = 90, hjust = 1, vjust = 0)) +
    xlab('Sample') +
    ylab('Counts')
}


ncm <- sweep(cm, 2, sf, '/')

print(plotColsums(cm))
```

Bar chart of counts after normalisation using estimated differenes in samples, primarily library size. Uses the [estimateSizeFactorsForMatrix](https://rdrr.io/bioc/ABSSeq/man/estimateSizeFactorsForMatrix.html) function from the DESeq2 package:


```{r normsum}
print(plotColsums(ncm))
```

Scatter plot of log10 counts for each of 20 genes (no particular selection). Overlaid violin plots and boxplots to illustrate distribution. Log10 counts from non-normalised counts matrix.

```{r dist}
plotDist <- function(mat){
  as.data.frame(mat)[1:20, ] %>% 
    rownames_to_column('gene') %>% 
    pivot_longer(2:last_col(), names_to = 'sample') %>%
    ggplot(aes(gene, log10(value))) + 
    geom_violin(aes(colour = gene)) + 
    geom_jitter(size = 0.5, alpha = 0.2) +
    geom_boxplot(alpha = 0.2) +
    theme_classic() +
    theme(axis.text.x = element_text(angle = 75, hjust = 1, vjust = 1, size = 7), 
          legend.position = 'None') +
    labs(x = 'Gene', y = 'Log10 counts')
}

print(plotDist(cm))
```

Scatter plot of log10 counts for the same genes post-normalisation. The inter-quartile range of sample log10 count values for each gene is broadly reduced. 

```{r normdist}
print(plotDist(ncm))
```

This document additionally creates a raw counts matrix data file (**raw_counts_matrix.RData**), and a normalised counts matrix data file (**norm_counts_matrix.RData**).